transformations/transformations/global_var_offload.py

```diff
@@ -0,0 +1,275 @@
+
+    def transform_subroutine(self, routine, **kwargs):
+        role = kwargs.get('role')
+        if role == 'driver':
+            self.process_driver(routine)
+        elif role == 'kernel':
+            self.process_kernel(routine, **kwargs)
+
+    def process_driver(self, routine):
+        """
+        Add offload and/or copy-back directives for the imported variables.
+        """
+
+        update_device = ()
+        update_host = ()
+
+        # build offload pragmas
+        if self._acc_copyin:
+            update_device += as_tuple(Pragma(keyword='acc',
+                                             content='update device(' + ','.join(self._acc_copyin) + ')'),)
+        if self._enter_data_copyin:
+            update_device += as_tuple(Pragma(keyword='acc',
+                                             content='enter data copyin(' + ','.join(self._enter_data_copyin) + ')'),)
+        if self._enter_data_create:
+            update_device += as_tuple(Pragma(keyword='acc',
+                                             content='enter data create(' + ','.join(self._enter_data_create) + ')'),)
+        if self._exit_data:
+            update_host += as_tuple(Pragma(keyword='acc',
+                                           content='exit data copyout(' + ','.join(self._exit_data) + ')'),)
+        if self._acc_copyout:
+            update_host += as_tuple(Pragma(keyword='acc', content='update self(' + ','.join(self._acc_copyout) + ')'),)
+
+        # replace Loki pragmas with acc data/update pragmas
+        pragma_map = {}
+        for pragma in FindNodes(Pragma).visit(routine.body):
+            if pragma.keyword == 'loki':
+                if 'update_device' in pragma.content:
+                    if update_device:
+                        pragma_map[pragma] = update_device
+                    else:
+                        pragma_map[pragma] = None
+
+                if 'update_host' in pragma.content:
+                    if update_host:
+                        pragma_map[pragma] = update_host
+                    else:
+                        pragma_map[pragma] = None
+
+        routine.body = Transformer(pragma_map).visit(routine.body)
+
+        # build new imports to add offloaded global vars to driver symbol table
+        new_import_map = {}
+        for s in self._var_set:
+            if s in routine.symbol_map:
+                continue
+
+            if new_import_map.get(self._modules[s], None):
+                new_import_map[self._modules[s]] += as_tuple(s)
+            else:
+                new_import_map.update({self._modules[s]: as_tuple(s)})
+
+        new_imports = ()
+        for k, v in new_import_map.items():
+            new_imports += as_tuple(Import(k, symbols=tuple(Variable(name=s, scope=routine) for s in v)))
+
+        # add new imports to driver subroutine sepcification
+        import_pos = 0
+        if (old_imports := FindNodes(Import).visit(routine.spec)):
```

Comment:
[No action required] is there a technical reason that we have to add these imports after existing imports instead of just prepending them to the `spec`?

Reply:
No technical reason. In that case I would also like to add a comment after the Loki generated imports, just to make a clear demarcation. Let me know please which implementation you would prefer.

Reply:
Thanks for explaining. Either is fine with me - @mlange, any preferences?
My question was motivated by the fact that I would have been lazy and therefore would have just prepended the imports, so was wondering if there was a specific reason.

Reply:
No preference, was just curious.